pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/KafkaLowLevelConsumerRoutingTableBuilder.java

```diff
@@ -0,0 +1,282 @@
+    // have a weight of 2, which is the maximum segment count minus the segment count for that replica. Thus, each
+    // replica other than the replica(s) with the maximum segment count should have a chance of getting a segment
+    // assigned to it. This corresponds to alternative three below:
+    //
+    // Alternative 1 (weight is sum of segment counts - segment count in that replica):
+    // (6 - 1) = 5 -> P(0.4166)
+    // (6 - 2) = 4 -> P(0.3333)
+    // (6 - 3) = 3 -> P(0.2500)
+    //
+    // Alternative 2 (weight is max of segment counts - segment count in that replica + 1):
+    // (3 - 1) + 1 = 3 -> P(0.5000)
+    // (3 - 2) + 1 = 2 -> P(0.3333)
+    // (3 - 3) + 1 = 1 -> P(0.1666)
+    //
+    // Alternative 3 (weight is max of segment counts - segment count in that replica):
+    // (3 - 1) = 2 -> P(0.6666)
+    // (3 - 2) = 1 -> P(0.3333)
+    // (3 - 3) = 0 -> P(0.0000)
+    //
+    // Of those three weighting alternatives, the third one has the smallest standard deviation of the number of
+    // segments assigned per replica, so it corresponds to the weighting strategy used for segment assignment. Empirical
+    // testing shows that for 20 segments and three replicas, the standard deviation of each alternative is respectively
+    // 2.112, 1.496 and 0.853.
+    //
+    // This algorithm works as follows:
+    // 1. Gather all segments and group them by Kafka partition, sorted by sequence number
+    // 2. Ensure that for each partition, we have at most one partition in consuming state
+    // 3. Sort all the segments to be used during assignment in ascending order of replicas
+    // 4. For each segment to be used during assignment, pick a random replica, weighted by the number of segments
+    //    assigned to each replica.
+
+    // 1. Gather all segments and group them by Kafka partition, sorted by sequence number
+    Map<String, SortedSet<String>> sortedSegmentsByKafkaPartition = new HashMap<String, SortedSet<String>>();
+    for (String helixPartitionName : externalView.getPartitionSet()) {
+      // Ignore segments that are not low level consumer segments
+      if (!SegmentNameBuilder.Realtime.isRealtimeV2Name(helixPartitionName)) {
+        continue;
+      }
+
+      String kafkaPartitionName = SegmentNameBuilder.Realtime.extractPartitionRange(helixPartitionName);
+      SortedSet<String> segmentsForPartition = sortedSegmentsByKafkaPartition.get(kafkaPartitionName);
+
+      // Create sorted set if necessary
+      if (segmentsForPartition == null) {
+        segmentsForPartition = new TreeSet<String>(new Comparator<String>() {
+          @Override
+          public int compare(String firstSegment, String secondSegment) {
+            // Sort based on sequence number, falling back on string comparison in case there is an exception
+            try {
+              int firstSegmentSequenceNumber =
```

Comment:
Sure, we can make that part of another PR. I'd like to avoid scope creep during review time though. :)